lib/ecto/changeset.ex

```diff
@@ -1877,6 +1877,10 @@ defmodule Ecto.Changeset do
     {validations, schema} =
       case changeset do
         %Ecto.Changeset{validations: validations, data: %schema{}} ->
+          unless function_exported?(schema, :__schema__, 1) do
+            raise ArgumentError, "unsafe_validate_unique/4 does not work with schemaless changesets"
+          end
+
           {validations, schema}
         %Ecto.Changeset{} ->
```

Comment:
Perhaps it is no longer worth having this clause? Check if it is a struct and if function_exported? is defined instead in a single conditional, without the case?

Reply:
Oh yeah good point, I will make that change. Btw this function won't work for embeds either right? Would it also be ok to add a check for the `__meta__` field and raise in its absence?

Reply:
I think pattern matching on __meta__ is enough, yeah!

Reply:
Updated the code. I think it has to be split like this otherwise someone can randomly put `__meta__` in a non-schema struct? But let me know if I'm being a newb.